### PR TITLE
[EMBR-1196] Create a retry queue for each endpoint.

### PR DESCRIPTION
**Pull request by lucaslabari**
## Goal

- Created a retry queue for each endpoint. 
- Now it loads and saves from cache the new map that holds the retry queues for each endpoint. 
- When loads return null, we still need to verify if it's returning null because of an exception caused by an old file version. (we can discuss doing this differently)
- Now, the limits are established for each endpoint.

## Testing

Added new unit tests and updated others.

### CODE REVIEW

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt:73]: This behavior is inconsistent with returning from the `sessions` collection higher up. Having said that, I'm unsure on whether we should bother changing the behavior to get deterministic request order as part of this PR. @bidetofevil thoughts?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt:53]: This works fine, ideally I'd like to change the strategy so the oldest retry gets dropped, rather than everything new. Also, we should ideally record somewhere that a retry was dropped. Not something to do in this PR

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceApiService.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceApiService.kt]: Agree, that's a better way. Done.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt]: Renamed, but added @SerializedName("cachedPayload") so it doesn't break compatibility with already existing files.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt]: Yeap, totally. Updated.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt]: I didn't use putIfAbsent because it's not what I needed for this case, but ended up using `getOrPut()` that is also a thread safe method.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt:227]: As you mentioned in another comment I wrapped the map inside `FailedApiCallsPerEndpoint` and improved the implementation of the add method, so we the `associateBy` wouldn't apply.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt]: I don't think so, what we want here is to poll the next item we want to retry, taking into account the priorities between types/endpoints.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/DeliveryFailedApiCalls.kt:90]: So, the idea we discussed with @bidetofevil was to prioritize sessions and then send payloads in the order they were enqued. 
Basically, we are adding the prioritization of sessions, because the order by queued time was already there as the previous implementation was a FIFO queue.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt:53]: Why are new attempts more valuable than the oldest ones?

Comment (fractalwrench) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt:53]: Good question - I don't think it's a clear cut distinction. I'd say they're slightly more valuable as folks care about things that happened recently. Also, there's less chance we're dealing with a bad scenario, like data corruption meaning an old session was stuck on a device & is unsendable.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt:15]: what about just calling it `endpoint()` since we're not transforming the object?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt:22]: Do we want to have a default value that isn't null so if we have a ApiRequest that doesn't fit this scheme later, it'll at least fall through to a generic queue? Something like `Endpoint.Unknown`?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt:101]: Can this just be returned by `FailedApiCallsPerEndpoint` instead of exposing the count?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt:24]: Do we need to expose this? Is it possible for the interface of the object to just be `add()` `getNext()`, `clear()`, and `hasAnyFailedApiCalls()`, leaving the limits, counts, etc. be internal to the object?

Basically, the retry manager should ask: "do you have a request for me to retry?" and this will just return if it has one. Not sure if it needs to expose more than that right now before we add the 429 stuff...

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (bidetofevil) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt:63]: probably need to synchronize this call with any methods that can modify the internal map? Given the amount of traffic we have, I don't think we need to be more clever than that the the locking

**Review by bidetofevil — COMMENTED** [review]
Looks good in general! Just some comments about naming, whether we need to expose all the methods that are currently public on `FailedApiCallsPerEndpoint`, and a couple of simplification/cleanup suggestions

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt:53]: @eliabsisay I created [this Story](https://www.notion.so/embraceio/DCI-1678d2876bc247a8951b7ee79f25afe2?p=c24c5a8772ec4089890531bd7a693d72&pm=s) to work on it.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt]: yeap, that's better, updated.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryCacheManager.kt:213]: We don't need to delete it because it will be overwritten the next time we save it. And deleting would be bad because we could be missing data if the requests retry don't succeed.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt]: Good catch, updated.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/api/EmbraceUrl.kt]: Yes, that's a good idea, I added `Endpoint.Unknown`

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceDeliveryRetryManager.kt]: Yes, I moved this to FailedApiCallsPerEndpoint.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt]: Done. I had to update a bit the code in the retryManager to avoid using the count. I leaved just the methods you are mentioning.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt:9]: I will do the rename in another PR so I don't complicate the review of this PR and also because we need to modify other components to be consistent on the naming we give to these Api Calls.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt
Comment (lucaslabari) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/FailedApiCallsPerEndpoint.kt]: As we talked on Slack, this class is accessed by a single thread, so I just added a comment in case we start accessing it from multiple threads.

**Review by bidetofevil — APPROVED** [review]
Looks good! Lets wait until the release is done before merging this?

**Pull request merged**
